components/tidb_query/src/codec/row/v2/decoder.rs

```diff
@@ -0,0 +1,332 @@
+use tikv_util::collections::HashMap;
+use tipb::ColumnInfo;
+
+pub trait RowDecoder: NumberDecoder + IntDecoder {
+    // | version | flag | number_of_not_null_values | number_of_null_values | not_null_value_ids | null_value_ids | value_offsets | values
+    fn decode_row(
+        &mut self,
+        ctx: &EvalContext,
+        cols: &HashMap<i64, ColumnInfo>,
+    ) -> Result<HashMap<i64, Datum>> {
+        let mut row = HashMap::with_capacity_and_hasher(cols.len(), Default::default());
+        if self.read_u8()? != super::CODEC_VERSION {
+            panic!("version not matched");
+        }
+        let is_large_id = self.read_u8()? == 1;
+
+        // read ids count
+        let not_null_cnt = self.read_u16_le()? as usize;
+        let null_cnt = self.read_u16_le()? as usize;
+
+        let ids: Vec<i64>;
+        let offsets: Vec<usize>;
+        if is_large_id {
+            ids = self.read_ints_le(not_null_cnt + null_cnt, u32::to_i64)?;
+            offsets = self.read_ints_le(not_null_cnt, u32::to_usize)?;
+        } else {
+            ids = self.read_ints_le(not_null_cnt + null_cnt, u8::to_i64)?;
+            offsets = self.read_ints_le(not_null_cnt, u16::to_usize)?;
+        };
+
+        let values = self.bytes();
+        let (not_null_ids, null_ids) = ids.split_at(not_null_cnt);
+        for (id, tp) in cols {
+            if let Ok(idx) = not_null_ids.binary_search(id) {
+                let offset = offsets
+                    .get(idx)
+                    .expect("should always have value in offsets");
+                let start = if idx > 0 { offsets[idx - 1] } else { 0usize };
+                row.insert(*id, (&values[start..*offset]).decode_datum(ctx, tp)?);
+            } else if null_ids.binary_search(id).is_ok() {
+                row.insert(*id, Datum::Null);
+            } else {
+                row.insert(*id, tp.get_default_val().decode_datum(ctx, tp)?);
+            }
+        }
+        Ok(row)
+    }
+}
+
+impl<T: BufferReader> RowDecoder for T {}
+
+trait DatumDecoder: NumberDecoder + DecimalDecoder + JsonDecoder {
+    #[inline]
+    fn decode_datum(&mut self, ctx: &EvalContext, ft: &dyn FieldTypeAccessor) -> Result<Datum> {
+        use core::convert::TryInto;
+        use tidb_query_datatype::FieldTypeTp::*;
+        match ft.tp() {
+            Tiny | Short | Long | LongLong | Int24 | Year => {
+                if ft.is_unsigned() {
+                    Ok(Datum::U64(self.decode_u64()))
+                } else {
+                    Ok(Datum::I64(self.decode_i64()))
+                }
+            }
+            VarChar | VarString | String | TinyBlob | MediumBlob | LongBlob | Blob => {
+                Ok(Datum::Bytes(self.bytes().to_vec()))
+            }
+            Double | Float => Ok(Datum::F64(f64::from_bits(self.decode_u64()))),
+            Date | DateTime | Timestamp => Time::from_packed_u64(
+                self.decode_u64(),
+                ft.tp().try_into().unwrap(),
```

Comment:
To make it compitable with the `prost`, prefer
```suggestion
                ft.accessor().tp().try_into().unwrap(),
```